### ISSUE TITLE
Feature - Rename File Dialog

### ISSUE THREAD

**Issue by antis81** [gitextensions/gitextensions#671]
Because we have to deal with case insensitive filenames under Windows and git won't notice case changes, a file rename dialog would be nice, that does the "git mv" and "git mv --force" command.

**Comment by CyDragon80** [comment]
Just had a similar situation. It would be nice if the right-click context menu on a file down in the File Tree view offered a 'Git Rename' where we could enter a new name and essentially do a "git mv -f FileName.ext filename.ext" in the background. One of the very few things I've had to leave the GUI to do.

**Comment by RussKie** [comment]
The file tree is not displaying the working directory content, it shows the snapshot of the repository at the commit.
File management lies outside of GE.

**Comment by gbburkhardt** [comment]
> 
> 
> The file tree is not displaying the working directory content, it shows the snapshot of the repository at the commit.
> File management lies outside of GE.

I think you've misunderstood the feature request.  I can't find any way in Git Extensions to do the equivalent of 'git mv'.  That's not "file management" - it's a legitimate git operation.  Whether it goes into the File Tree View context menu is another issue.  For me, that would be a natural place.

**Comment by mstv** [comment]
I agree with both of you. In the meantime, the situation has changed a litte since we have the virtual commit(s) "working directory" (and "commit index").
When at "working directory", the file tree should have a context menu item for renaming files. There should be an easy way to rename the file(s) to lower case, which is a quite common use case IMO.

**Comment by mstv** [comment]
Though the file tree is hidden for the virtual commits.

**Comment by gbburkhardt** [comment]
There ought to be a way in GitExtensions to perform a "git mv".  I don't care where it is.

**Comment by gerhardol** [comment]
>  I don't care where it is.

There is no natural place to base this information. The addition will be difficult to use as I see it.

> Though the file tree is hidden for the virtual commits.

The file tree is read with git-ls-tree, not available for the worktree or index. It is possible to build the file tree from git-ls-files, but it would be hacky, need to rebuilt and hard to handle files handled by git and those that are not.

It is reasonable to do a mv on worktree diff, but that would require that files are modified before they can be renamed which is weird.

The natural way is to add this to any revision but make the change in the worktree/index.
A few such actions exists already, this is even weirder.

![image](https://user-images.githubusercontent.com/6248932/56608205-86df5480-660a-11e9-9905-b4064a690435.png)

**Comment by RussKie** [comment]
> There should be an easy way to rename the file(s) to lower case, which is a quite common use case IMO.

👍 definitely. It is a PIA to do in Windows

> > Though the file tree is hidden for the virtual commits.
> 
> The file tree is read with git-ls-tree, not available for the worktree or index. It is possible to build the file tree from git-ls-files, but it would be hacky, need to rebuilt and hard to handle files handled by git and those that are not.

I don't think it would be too difficult to list files and folders from a HDD.

**Comment by mstv** [comment]
The file tree of the HEAD commit is the one which contains the files that can be renamed in Git (except the files which have been added in the virtual commits which can be ignored IMO).
Hence this menu item could be added there. Usually, I rather disable than hide menu items. In this case I would hide it for non-HEAD commits.

**Comment by gerhardol** [comment]
> The file tree of the HEAD commit is the one which contains the files that can be renamed in Git (except the files which have been added in the virtual commits, which can be ignored IMO).
Hence this menu item could be added there. Usually, I rather disable than hide menu items. In this case, I would hide it for non-HEAD commits.

git-rename operates on files tracked by git in working tree. HEAD and index cannot be ignored IMO, it is just almost correct.
So rename and similar should operate on the working tree. If the virtual commit "Working directory" had a file tree, it could be included there.

There could be a menu item "Show in working directory" or something for normal commits (both diff and filetree tabs) to move to the tab. (This will be similar to Show in Folder with context menu alowwing to git rename etc).

**Comment by NN---** [comment]
It would be nice to have TortoiseGit rename feature.
![](https://tortoisegit.org/docs/tortoisegit/images/ContextMenuFileControl.png)

**Comment by mstv** [comment]
Yes, @NN---. To implement it as part of the shell extension should be quite easy.
(It can be reached from within GE by means of the context menu item `Show in folder`.)

**Comment by gerhardol** [comment]
Two menu actions for tracked files:
* Shell extension
* Worktree virtual commit

**Comment by mstv** [comment]
This feature request slipped down the list.
For the time being, a `cmd` script can be used to lower case all files in the current folder.
(I know it could be done more elegant using PS but for execution policies.)
```
@echo off
setlocal enabledelayedexpansion
for %%f in (*) do (
    set file=%%f
    for %%b in (a b c d e f g h i j k l m n o p q r s t u v w x y z) do set "file=!file:%%b=%%b!"
    if not %%f == !file! (
        echo %%f - !file!
        git mv %%f !file!
    )
)
```
